### DISCUSSION TITLE
Automatic detection of ExceptionList.txt based patch and switching to FecContainer.fec based solution

### DISCUSSION THREAD

**Discussion by otzi81** [Mr-MIBonk/M.I.B._More-Incredible-Bash#182] [Ideas]
Hello guys,

My idea is what I mentioned in the title.

Why I am thinking about it (sorry for so long description)?

My problem is that my unit's software was updated to 1326 (A3 car) and has been patched by one guy in the mid of 2021.
He set then maps expiration date to be valid until 2022, however I asked for the lifetime. I could do this myself, but I did not have much time for reading then, so I decided to pay for this.

After some time I realized that FEC code for maps is set to 0230002c. Recently I discovered your project and wanted to use it to improve maps validity date :). I do not want to pay him again for something that he should already do. I went through the Wiki and spot this problem description:
https://github.com/Mr-MIBonk/M.I.B._More-Incredible-Bash/wiki/ExceptionList-BUG

If I got it right, this issue will appear right after installing M.I.B., so I will not have possibility to run the option of M.I.B. that allows to set root password (under config menu), as I guess the reboot after 60 seconds will happen before green menu would be available.

Unfortunately I am not sure which patching method this guy used. I enabled green menu via OBD11, formatted SD card with exFAT system (FAT32 is not available as the SD card's capacity is probably to big for it). Then I put the card into SD slot one in the MIB, launched green menu, but did not find menu from M.I.B. project. Thus I suspect that the guy had to use ExceptionList patch method, or uninstalled M.I.B. project from my unit, and disabled green menu. I cannot be sure. Anyway, I tried to login to my unit, using the password I found for my SW version in the internet but it does not match. I thought that maybe the guy, who patched my MIB changed the pass, so I contacted him, but he refused to tell me the password. Without knowing the root password I do not want to risk M.I.B. installation, as if indeed ExceptionList patch was used, then I will not be able to login to the unit via ssh. I am not sure if UART login requires providing password or not, but for me this is not an option, as I do cannot pull it out, to get access to the connectors.

What I also did is I found files with 1326 SW image in the internet, using dumpifs tool I extracted it and got the hash of the root password. After googling this hash it turned out that it matches different password, that I found in the internet for 1326 software. If the guy used same image, maybe this password will work. I am going to check it, but I do not have big hope that it will be fine.

But after my long story, let' s go to the point.

Would it be possible to add a functionality, that would let's say, detect 2 or 3 consecutive unit restarts, each after ~60 seconds? If such situation is detected, that most probably means that we have a problem with installing M.I.B. on the top of ExceptionList patched unit. Then call _rm /mnt/efs-persist/FEC/mmx_fec_ids.bin_ to remove the file. After that M.I.B. could be used to FecContainer.fec based patch. I guess it should be possible if any part of M.I.B. source code is executed automatically during unit startup. I not, then I am afraid nothing can be done.

But then, maybe it would be possible to reset the root's password to some default e.g. "mibproject123" by the installer during the M.I.B. installation?

Guys, please let me know what do you think about it? Maybe I will not be the only person, who would like to install M.I.B. but has already patched unit with unknown method and does not know the root password. From my side, once I find some time, I will try to understand how exactly M.I.B. is implemented to bring some effort from my side for the project.. After cloning the repo I see that mainly there are a lot of shell scripts, so getting familiar with it will require some time.

Thanks!

**Comment by harman-f** [comment]
No, the bug will not appear directly after M.I.B install.

If EL method was used to aptch your unit, this bug might appear at any time of a reboot.

Install M.I.B and run addFec function.
It will revert EL to stock file and adjust your FecContainer.

**Reply by otzi81** [reply]
Hi @harman-f 

Thanks a lot for your prompt answer!

Just to confirm that my understanding is correct:

1. EL issue is independent on the M.I.B. at all.
2. If my unit has been patched with EL method, it already could suffer from ExceptionList bug.
3. If I have never observed restarts that means either my unit is immune to the EL bug issue or it has been patched with some other method.
4. Installation of M.I.B. will not cause that EL bug will start to appear on my unit. 

As I do not know how the unit has been patched, but for sure FEC code 02300xx was updated to 0230002c, so some patch had to be applied, I would like to confirm the steps to perform.

Are the two steps mentioned by you sufficient in my case?
"Install M.I.B and run addFec function.
It will revert EL to stock file and adjust your FecContainer."

Just to remind, I do not know the method used to patch my unit and I do not have root password, so  in case something is wrong and integration into the file system is needed, I stuck.

Or maybe I could perform the steps like:
1. Install M.I.B. 3.5
2. Go to =>m.i.b.<= and use "Create backup of your unit and keep it safe"
3. Go to =>m.i.b.<=/config and use "Set root password to harman_f", to set root's password to something known to me
4. Go to =>m.i.b.<= and use "patch_ifs_root_aio". This should set FEC for map validity to lifetime. CarPlay/Android auto are in my car from the factory, so they should remain.

What do you suggest?

Thank you in advance for your feedback and have a nice day.

edit:
I did as you suggested and it worked. In the FEC codes I can see my old one for 0230002c but new one 023000ee was added at the end. I did not try to install new maps yet, but I think it should be ok. Also I used option to set known root password to get control back on my unit, just in case :)

Thanks a lot for your help!

**Reply by surfer74** [reply]
> edit: I did as you suggested and it worked. In the FEC codes I can see my old one for 0230002c but new one 023000ee was added at the end. I did not try to install new maps yet, but I think it should be ok. Also I used option to set known root password to get control back on my unit, just in case :)

Hello [otzi81](https://github.com/otzi81) 
It`s look loki a have the same issues as you - my unit was patched some years ago for money with USB modem and now I cant patch it because it "already patched" 
My FEC now 02300030. I made addFec function as described below and during this process saw two FECs - my old 02300030 and a new 023000ee. Via RED menu I see only the old 02300030.
Will new 023000ee work? Did you try to install new maps?
Thank you and thank of course to [harman-f](https://github.com/harman-f) for answer below andthis project

**Reply by otzi81** [reply]
Hi @surfer74 

In the red menu I can see both FEC codes, the old one and the new one 023000ee that is at the end of the FEC codes list.
It is strange, that you can see only the old one. Did you check whole the list, until the last entry? My code, as I mentioned, is the very last one.

Regarding maps update, yes, I was able to successfully update it to the version 2023 few day ago. In the meantime I noticed that 2023/24 maps have been released, so I have already prepared SD card and I am going to update it again while I will have some longer drive. Whole update process took for me ~40 minutes.

**Reply by surfer74** [reply]
> is the very last one.

Bingo, I scrolled till the end and really found 023000ee!
Thank you so much  
[otzi81](https://github.com/otzi81)
Today or tomorrow I will try to set a new maps

**Comment by Fanaticjonathanjordan** [comment]
Help!!! I tried to patch with EL and my M.I.B is in VW logo I connect it to uart with the 2 usb and I don't know how to enter the code mnt/efs-persist/FEC/mmx_fec_ids.bin recovery for the next restart please help I'm from Chile 🥺🥺
